fix: fix test_html_tags_removal

mubbsharanwar opened the pull request:
## Description
This PR addresses a test fix to encode `"` for  Django 5.2.

Django escapes `"` to `&amp;quot;` starting in 5.0.10 as part of security hardening, because leaving raw quotes in truncated HTML could allow attribute-breaking XSS payloads. This was fixed in [5.0.10](https://docs.djangoproject.com/en/5.0/releases/5.0.10/) by tightening strip_tags() and Truncator, which rely on a custom HTMLTruncator subclass of Python’s HTMLParser to safely walk and escape HTML content.

Issue: [37198](https://github.com/openedx/edx-platform/issues/37198)

mubbsharanwar committed:
fix: fix test_html_tags_removal

unescape "" from the html body.

issue: 37198

mfarhan943 reviewed (approved):
LGTM 👍

awais786 commented:
@mubbsharanwar I will review this also so wait for my approval.

awais786 commented on lms/djangoapps/discussion/rest_api/discussions_notifications.py:
```diff
@@ -447,7 +448,10 @@ def clean_thread_html_body(html_body):
     Get post body with tags removed and limited to 500 characters
     """
-    html_body = BeautifulSoup(Truncator(html_body).chars(500, html=True), 'html.parser')
+    truncated_body = Truncator(html_body).chars(500, html=True)
```
```
truncated_body = Truncator(html_body).chars(500, html=True)

# Django 5+ escapes quotes, so unescape them back for consistency
if django.VERSION >= (5, 0):
    truncated_body = html.unescape(truncated_body)

html_body = BeautifulSoup(truncated_body, 'html.parser')
```

awais786 replied:
also `Truncator(html_body).chars(500, html=True)` it as 2 occurances in code please fix there also.

mubbsharanwar replied:
- The `unescape` is working fine with 4.2 and 5.2, so without adding if condition, we can fix it for both versions. We have already added TODO to remove `escape` once upgraded to 5.2 for consistency.   

- The `unescape` is already used [there](https://github.com/openedx/edx-platform/blob/077e4ee4afe9627056e884b879506f8dacc274ec/cms/djangoapps/contentstore/utils.py#L2298) to handle the encoded html body, so we don't need to handle it.

hunzlahmalik commented on lms/djangoapps/discussion/rest_api/discussions_notifications.py:
```diff
@@ -447,7 +448,10 @@ def clean_thread_html_body(html_body):
-    html_body = BeautifulSoup(Truncator(html_body).chars(500, html=True), 'html.parser')
+    truncated_body = Truncator(html_body).chars(500, html=True)
+    # TODO: Remove unescape once upgraded to Django 5 and fix test_html_tags_removal by replacing " with &quot;
+    truncated_body = html.unescape(truncated_body)
```
Is there a way to not use unescape?
Looking into why `html.escape` is becoming problem. The way I see it that the testcase failed with a difference in `script` tag.

hunzlahmalik replied:
Why are we putting it off for the future?

mubbsharanwar replied:
There is a way not to use `unescape`, for that we need to update the test `expected_output` based on the Django version.
Due to a security fix(mentioned in description), replaced `"` with `&quot;` in Django 5.0, we should keep our implementation according to the upgraded version and update `expected_output` once upgraded to 5.2 as mentioned in TODO.

hunzlahmalik replied:
Yes, can we do this way, add if based on Django version in the test.

Plus can you also check what will be the effect of keeping `&quot;` instead of `"`.

mubbsharanwar commented on lms/djangoapps/discussion/rest_api/tests/test_discussions_notifications.py:
```diff
@@ -108,11 +109,14 @@ def test_html_tags_removal(self):
-                           '<p style="margin: 0">Embedded video: </p>'
-                           '<p style="margin: 0">Script test: alert("hello");</p>'
-                           '<p style="margin: 0">Some other content that should remain.</p>')
+        excepted_script_quot = 'alert(&amp;quot;hello&amp;quot;);' if django.VERSION >= (5, 0) else 'alert("hello");'
```
In Django 5.0, `"` encoded to `&amp;quot;` to fix a security hardening, so we updated the expected_output, which contains `&amp;quot;` instead of `"`  in 5.2 that would be render correctly in the browser.

awais786 commented:
@kdmccormick please review.

awais786 commented:
@kdmccormick Please check this.

awais786 commented:
@feanil Please review.

awais786 committed:
Merge branch 'master' into mubbsharanwar/issue-37198

awais786 commented:
@mubbsharanwar Last time you made changes in this method [clean_thread_html_body](https://github.com/openedx/edx-platform/blob/ec83b2aa647c0d2527e8e3a28f43d43faa5d6f65/lms/djangoapps/discussion/rest_api/discussions_notifications.py#L446)

mubbsharanwar commented:
> @mubbsharanwar Last time you made changes in this method [clean_thread_html_body](https://github.com/openedx/edx-platform/blob/ec83b2aa647c0d2527e8e3a28f43d43faa5d6f65/lms/djangoapps/discussion/rest_api/discussions_notifications.py#L446)

Yes, we used `unescape`, but then we removed it and fixed the test only.
